### PR TITLE
feat(aws): Allow user supplied user data to override user provider user data (typically, udf templates)

### PR DESCRIPTION
**Pull request by jonsie**
- Additionally allows user supplied user data to use the standard UDF tokens
- Adds parameter `userDataOverride ` to the `BasicAmazonDeployDescription` and `ModifyAsgLaunchConfigurationDescription`
-  Moves UDF tokenization code out to util `UserDataTokenizer`
- Consolidates the user data aggregation logic via `UserDataProviderAggregator`
- Cleans up `UserDataProvider` (requires a few small -nflx changes)
- Introduces spinnaker extension point `UserDataTokenizer`.

### CODE REVIEW

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregator.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  /**
+   * Aggregates all user data. First iterates through all providers and joins user data with a
+   * newline. Then, adds the user supplied base64 encoded user data and again joins with a newline.
+   * The result is such that the user supplied base64 encoded user data is always appended last to
+   * the user data.
+   *
+   * <p>Note, if the {@link UserDataProvider.UserDataRequest} parameter overrideDefaultUserData is
+   * true, then the user data from the providers is skipped and the user supplied base64 encoded
+   * user data is used as the override. If this is the case, the standard set of user data format
+   * tokens ("%%app%%, for example) are replaced in the user data - effectively processing the user
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregator.java:33]:
```suggestion
   * tokens ("%%app%%", for example) are replaced in the user data - effectively processing the user
```

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregator.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+   * data as a UDF template.
+   *
+   * @param userDataRequest {@link UserDataProvider.UserDataRequest}
+   * @return String
+   */
+  public String aggregate(UserDataProvider.UserDataRequest userDataRequest) {
+    List<String> allUserData = new ArrayList<>();
+    if (providers != null && (!userDataRequest.isOverrideDefaultUserData())) {
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregator.java:41]:
```suggestion
    if (providers != null && !userDataRequest.isOverrideDefaultUserData()) {
```

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            .replace("%%countries%%", countries)
+            .replace("%%devPhase%%", devPhase)
+            .replace("%%hardware%%", hardware)
+            .replace("%%zone%%", zone)
+            .replace("%%cluster%%", cluster)
+            .replace("%%stack%%", stack)
+            .replace("%%detail%%", detail)
+            .replace("%%tier%%", "");
+
+    if (userDataRequest.getLaunchTemplate() != null && userDataRequest.getLaunchTemplate()) {
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java:55]:
```suggestion
    if (userDataRequest.getLaunchTemplate() != null) {
```

Odd conditional, isn't the second expression is redundant?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+          result
+              .replace("%%launchconfig%%", userDataRequest.getLaunchSettingName())
+              .replace("%%launchtemplate%%", "");
+    }
+
+    return result;
+  }
+
+  private static boolean isPresent(String value) {
+    return value != null && !value.isBlank();
```
Comment (robzienert) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java:71]: nit, but we could use `!Strings.isNullOrEmpty()` here.

File: clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregatorSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,105 @@
+package com.netflix.spinnaker.clouddriver.aws.deploy.userdata
+
+import spock.lang.Specification
+
+class UserDataProviderAggregatorSpec extends Specification {
```
Comment (robzienert) [inline clouddriver-aws/src/test/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregatorSpec.groovy:5]: nit: Would prefer to see the base64-encoded test inputs as test resources instead and then base64 the file contents inside the test class, that way if the test inputs need to change, it'll be more readily obvious what specific part changed. Further, we won't need to worry about these comments drifting from the actual values.

**Review by robzienert — APPROVED** [review]

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregator.java (hunk 2 above)
Comment (jonsie) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataProviderAggregator.java:41]: Weird.  Must've been an artifact from some change I made and then deleted.  Nice eye.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java (hunk 3 above)
Comment (jonsie) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java:55]: In this case `getLaunchTemplate` is a `Boolean` object which can be null and we only want to run this code if it's both not null and true.

**Review by cfieber — APPROVED** [review]
nice cleanup, a couple minor suggestions

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package com.netflix.spinnaker.clouddriver.aws.deploy.userdata;
+
+import com.netflix.frigga.Names;
+
+/** Utility class to replace tokens in user data templates. */
+public class UserDataTokenizer {
```
[the remark this answers is not in the document]
Comment (jonsie) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java:6]: I like that.  Making this an extension point could potentially greatly simplify the use case too!

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java (hunk 3 above)
Comment (robzienert) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java:55]: I didn't realize what value it was. Poor method / property naming, that one. Oh well. Thanks!

Comment (jonsie) [reply clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/userdata/UserDataTokenizer.java:55]: Yeah that's the default name you get from lombok when you use the `Boolean` object.  👎

File: clouddriver-api/clouddriver-api.gradle [hunk 7]
```diff
@@ -23,6 +23,7 @@ dependencies {
   api("com.netflix.spinnaker.kork:kork-plugins-api")
   api("com.netflix.spinnaker.kork:kork-credentials-api")
 
+  implementation "com.netflix.frigga:frigga"
```
Comment (jonsie) [inline clouddriver-api/clouddriver-api.gradle:26]: I think it's ok to add frigga to clouddriver-api - it is lightweight (believe it is just Java) and stable.

**Review by cfieber — APPROVED** [review]

**Pull request merged**